### PR TITLE
Aspire fixes

### PR DESCRIPTION
**Pull request by tmat**
Fixes various issues found while testing Aspire.

Fixes https://github.com/dotnet/sdk/issues/44044

### CODE REVIEW

**Comment by tmat** [comment]
@phil-allen-msft ptal

File: src/BuiltInTools/dotnet-watch/Browser/BrowserConnector.cs [hunk 1]
```diff
@@ -65,8 +66,12 @@ await Task.WhenAll(serversToDispose.Select(async server =>
                 }
             }
 
-            // Attach trigger to the process that launches browser on URL found in the process output:
-            processSpec.OnOutput += GetBrowserLaunchTrigger(projectNode, projectOptions, server, cancellationToken);
+            // Attach trigger to the process that launches browser on URL found in the process output.
+            // Only do so for root projects, not for child processes.
+            if (projectOptions.IsRootProject)
```
Comment (davidfowl) [inline src/BuiltInTools/dotnet-watch/Browser/BrowserConnector.cs:71]: 👍🏾

File: src/BuiltInTools/dotnet-watch/DotNetWatcher.cs [hunk 2]
```diff
@@ -54,6 +39,21 @@ public override async Task WatchAsync(CancellationToken cancellationToken)
...
+                ProjectGraphNode? projectRootNode;
+                if (evaluationResult.ProjectGraph != null)
+                {
+                    projectRootNode = evaluationResult.ProjectGraph.GraphRoots.Single();
+                    var projectMap = new ProjectNodeMap(evaluationResult.ProjectGraph, Context.Reporter);
+                    staticFileHandler = new StaticFileHandler(Context.Reporter, projectMap, browserConnector);
+                }
+                else
+                {
+                    Context.Reporter.Verbose("Unable to determine if this project is a webapp.");
```
Comment (davidfowl) [inline src/BuiltInTools/dotnet-watch/DotNetWatcher.cs:52]: What's this condition? Why is it tied to web apps?

Comment (tmat) [reply src/BuiltInTools/dotnet-watch/DotNetWatcher.cs:52]: This code path is legacy mode (not Hot Reload). If we can't load the project graph we can't detect project capabilities and can't do browser refresh.

File: test/Microsoft.WebTools.AspireService.Tests/AspireServerServiceTests.cs [hunk 3]
```diff
@@ -352,53 +351,37 @@ private async Task listenForSessionUpdatesAsync(AspireServerService aspireServer
...
+
+            var notification = JsonSerializer.Deserialize<SessionNotification>(message, AspireServerService.JsonSerializerOptions);
+            Assert.NotNull(notification);
+
+            SessionNotification value = notification.NotificationType switch
+            {
+                NotificationType.ProcessRestarted => JsonSerializer.Deserialize<ProcessRestartedNotification>(message, AspireServerService.JsonSerializerOptions),
+                NotificationType.SessionTerminated => JsonSerializer.Deserialize<SessionTerminatedNotification>(message, AspireServerService.JsonSerializerOptions),
+                NotificationType.ServiceLogs => JsonSerializer.Deserialize<ServiceLogsNotification>(message, AspireServerService.JsonSerializerOptions),
+                _ => throw new InvalidOperationException($"Unexpected {notification.NotificationType}")
```
Comment (davidfowl) [inline test/Microsoft.WebTools.AspireService.Tests/AspireServerServiceTests.cs:380]: Maybe this shouldn't crash watch? It should instead log an error message but keep working?

Comment (davidfowl) [reply test/Microsoft.WebTools.AspireService.Tests/AspireServerServiceTests.cs:380]: Oh wait this is a test 😄

File: src/BuiltInTools/dotnet-watch/Aspire/AspireServiceFactory.cs [hunk 4]
```diff
@@ -59,21 +59,18 @@ public async ValueTask<string> StartProjectAsync(string dcpId, ProjectLaunchRequ
...
 
             RunningProject? runningProject;
             lock (_guard)
             {
-                if (!_sessions.TryGetValue(sessionId, out runningProject))
-                {
-                    return false;
-                }
+                runningProject = _sessions[sessionId];
+                _sessions.Remove(sessionId);
```
Comment (davidfowl) [inline src/BuiltInTools/dotnet-watch/Aspire/AspireServiceFactory.cs:70]: Can we make this robust against double deletes?

File: test/Microsoft.WebTools.AspireService.Tests/AspireServerServiceTests.cs (hunk 3 above)
Comment (tmat) [reply test/Microsoft.WebTools.AspireService.Tests/AspireServerServiceTests.cs:380]: Crash is better here. This should never happen.

File: src/BuiltInTools/dotnet-watch/Aspire/AspireServiceFactory.cs (hunk 4 above)
Comment (tmat) [reply src/BuiltInTools/dotnet-watch/Aspire/AspireServiceFactory.cs:70]: Wouldn't that mask bugs?

Comment (tmat) [reply src/BuiltInTools/dotnet-watch/Aspire/AspireServiceFactory.cs:70]: BTW, I'm working on more changes in this type.

**Review by davidfowl — APPROVED** [review]

**Review by davidfowl — APPROVED** [review]

**Pull request merged**